### ISSUE TITLE
[jit] warnings.warn is always printed

### ISSUE THREAD

**Issue by tunz** [pytorch/pytorch#15119]
## To Reproduce

```python
import torch
import torch.nn as nn


class M(torch.jit.ScriptModule):

    def __init__(self):
        super(M, self).__init__()
        self.softmax = nn.Softmax(dim=0)

    @torch.jit.script_method
    def forward(self, v):
        return self.softmax(v)

i = torch.Tensor(2)
m = M()
o = m(i)
```

```
/home/tunz/anaconda3/envs/torchdev/lib/python3.7/site-packages/torch/nn/modules/module.py:489: RuntimeWarning: Format options are not supported.
  result = self.forward(*input, **kwargs)
/home/tunz/anaconda3/envs/torchdev/lib/python3.7/site-packages/torch/nn/modules/module.py:489: RuntimeWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  result = self.forward(*input, **kwargs)
```

## Expected behavior

no warnings

## Environment
```
PyTorch version: 1.0.0a0+479481b
Is debug build: No
CUDA used to build PyTorch: 9.2.148

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.12.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.2.148
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 410.73
cuDNN version: Probably one of the following:
/usr/local/cuda-9.2/lib64/libcudnn.so
/usr/local/cuda-9.2/lib64/libcudnn.so.7
/usr/local/cuda-9.2/lib64/libcudnn.so.7.0.3
/usr/local/cuda-9.2/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] blas                      1.0                         mkl
[conda] magma-cuda92              2.4.0                         1    pytorch
[conda] mkl                       2018.0.3                      1
[conda] mkl-include               2019.1                      144
[conda] mkl_fft                   1.0.6            py37h7dd41cf_0
[conda] mkl_random                1.0.1            py37h4414c95_1
[conda] torch                     1.0.0a0+db5d313           <pip>
[conda] torch                     1.0.0a0+479481b           <pip>
```
## Additional context

A warning about `dim` parameter is always printed even if `dim` is not None.
https://github.com/pytorch/pytorch/blob/00a4c8d41cf8a7504a4300b3fc741d8a4ccc5795/torch/nn/functional.py#L1183-L1184
https://github.com/pytorch/pytorch/blob/00a4c8d41cf8a7504a4300b3fc741d8a4ccc5795/torch/nn/functional.py#L1114-L1118

After some debugging, I found that it's actually taking correct paths that don't have warnings. it's just printing every warnings at the start of each function regardless of actual paths. I guess some jit optimization pass is reordering `at::warn` op.

**Comment by zou3519** [comment]
Minimal repro:
```
import torch
import warnings

@torch.jit.script
def fn(x):
  if bool(x > 0):
    warnings.warn("foo")
  return x

fn(torch.tensor(0))
```
warns "RuntimeWarning: foo"

**Comment by zou3519** [comment]
I think constant propagation is running the `aten::warn` op... looking into this some more

**Comment by qbx2** [comment]
When I run:
```python
import torch.nn.functional as F


@torch.jit.script
def test(input, label):
    return F.cross_entropy(input, label)


x = torch.randn(5, 5)
y = torch.randint(size=(5,), high=5, dtype=torch.long)
print(test(x, y))
```

I get:
```$ python test_jit.py
/home/qbx2/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py:489: RuntimeWarning: Format options are not supported.
  result = self.forward(*input, **kwargs)
/home/qbx2/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py:489: RuntimeWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
  result = self.forward(*input, **kwargs)
/home/qbx2/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py:489: RuntimeWarning: reduction='elementwise_mean' is deprecated, please use reduction='mean' instead.
  result = self.forward(*input, **kwargs)
tensor(1.9669)
```

Is it related to this issue and ok to ignore them?
